{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b8486df-68c4-49a5-8527-bdf72d7975e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current work directory: C:\\Users\\yyyyx\\Work\\OC43_machine_learning\\Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import mygene\n",
    "import pandas as pd\n",
    "\n",
    "if os.getcwd().endswith(\"notebooks\"):\n",
    "    os.chdir(\"..\")\n",
    "\n",
    "sys.path.append(os.path.abspath(\"./\"))\n",
    "\n",
    "print(f\"Current work directory: {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e48cae01-a36d-4531-9f86-91429ef21e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_table('./Data/Bulk-seq_data/GSE238079_HCoV_nae_PAPER_raw_gene_counts.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "925b85ee-7665-4cd5-ae5e-769ad8b44769",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Input sequence provided is already in string format. No operation performed\n",
      "Input sequence provided is already in string format. No operation performed\n",
      "11 input query terms found dup hits:\t[('ENSG00000175711', 2), ('ENSG00000188660', 2), ('ENSG00000215156', 2), ('ENSG00000226519', 2), ('E\n",
      "368 input query terms found no hit:\t['ENSG00000002079', 'ENSG00000112096', 'ENSG00000139656', 'ENSG00000146521', 'ENSG00000148362', 'ENS\n"
     ]
    }
   ],
   "source": [
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "ensembl_ids = df['Gene'].unique().tolist()\n",
    "\n",
    "# Ensembl gene ID â†’ Gene symbol\n",
    "res = mg.querymany(ensembl_ids,\n",
    "                   scopes='ensembl.gene',\n",
    "                   fields='symbol',\n",
    "                   species='human')\n",
    "\n",
    "id2symbol = {\n",
    "    r['query']: r.get('symbol')\n",
    "    for r in res\n",
    "    if not r.get('notfound', False)\n",
    "}\n",
    "\n",
    "df['Gene_symbol'] = df['Gene'].map(id2symbol)\n",
    "df = df.loc[:, ~df.columns.str.contains('229E')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b15657c-0d56-40fc-a854-53bba55d578b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Gene_symbol   A_Mock   B_Mock A_OC43_24h B_OC43_24h A_OC43_72h B_OC43_72h\n",
      "0      TSPAN6  6004.21  5144.02    2413.02       2076        631       2429\n",
      "1        DPM1     1261  1190.61    625.577    686.011        554    736.387\n",
      "2       SCYL3  997.424  926.088    788.292    727.245    178.075    460.184\n",
      "3       FIRRM  111.576  749.151    106.708    128.755    309.248    101.816\n",
      "4         FGR       12       20         15         18         17         35\n"
     ]
    }
   ],
   "source": [
    "df = df.dropna(subset=['Gene_symbol'])\n",
    "df = df.drop(columns=['Gene'])\n",
    "\n",
    "cols = ['Gene_symbol'] + [col for col in df.columns if col != 'Gene_symbol']\n",
    "df = df[cols]\n",
    "\n",
    "# Using OC43 data only\n",
    "df = df.loc[:, ~df.columns.str.contains('229E')]\n",
    "\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d7a8dd6-054d-47e2-8270-b3c30e158f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"./CSV/Bulk-seq/GSE238079_converted.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b8568bc9-bb71-44c6-9703-fcf7f1df3c8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "mygene              3.2.2\n",
       "pandas              3.0.0\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "anyio               NA\n",
       "asttokens           NA\n",
       "biothings_client    0.4.1\n",
       "bottleneck          1.4.2\n",
       "brotlicffi          1.0.9.2\n",
       "certifi             2025.10.05\n",
       "click               8.2.1\n",
       "colorama            0.4.6\n",
       "comm                0.2.1\n",
       "cython_runtime      NA\n",
       "dateutil            2.9.0.post0\n",
       "debugpy             1.8.11\n",
       "decorator           5.1.1\n",
       "executing           0.8.3\n",
       "h11                 0.16.0\n",
       "httpcore            1.0.9\n",
       "httpx               0.28.1\n",
       "idna                3.7\n",
       "ipykernel           6.29.5\n",
       "jedi                0.19.2\n",
       "nt                  NA\n",
       "numexpr             2.11.0\n",
       "numpy               2.1.3\n",
       "packaging           24.2\n",
       "parso               0.8.4\n",
       "platformdirs        3.10.0\n",
       "prompt_toolkit      3.0.43\n",
       "psutil              5.9.0\n",
       "pure_eval           0.2.2\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              3.2.3\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.15.1\n",
       "pytz                2024.1\n",
       "six                 1.17.0\n",
       "stack_data          0.2.0\n",
       "tornado             6.4.2\n",
       "traitlets           5.14.3\n",
       "typing_extensions   NA\n",
       "wcwidth             0.2.5\n",
       "zmq                 26.2.0\n",
       "zoneinfo            NA\n",
       "zope                NA\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.30.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.7.2\n",
       "-----\n",
       "Python 3.11.11 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:34:19) [MSC v.1929 64 bit (AMD64)]\n",
       "Windows-10-10.0.26200-SP0\n",
       "-----\n",
       "Session information updated at 2026-02-13 15:41\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "\n",
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc_env",
   "language": "python",
   "name": "sc_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
