{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bbaa9706-ab3f-4202-b73b-c6344d312bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "  library(AnnotationDbi)\n",
    "  library(org.Hs.eg.db)\n",
    "  library(clusterProfiler)\n",
    "  library(enrichplot)\n",
    "  library(ggplot2)\n",
    "  library(tools)\n",
    "  library(here)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3b9fd89-7db0-4175-a82e-8ce66bd1de74",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ded4104e-ec59-4f1b-b18b-a9cc9094ea57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning'"
      ],
      "text/latex": [
       "'C:/Users/yyyyx/Work/OC43\\_machine\\_learning/Regression\\_Modeling\\_of\\_OC43\\_Coronavirus\\_Viral\\_Burden\\_from\\_Host\\_Single-cell\\_Transcriptomes\\_Using\\_Explainable\\_Machine\\_Learning'"
      ],
      "text/markdown": [
       "'C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning'"
      ],
      "text/plain": [
       "[1] \"C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "914ac7ea-0a27-4c78-9aa1-0a1d2062e901",
   "metadata": {},
   "outputs": [],
   "source": [
    "utils <- new.env()\n",
    "\n",
    "# 2. 그 공간 안에 스크립트 내용을 실행합니다.\n",
    "sys.source(here::here(\"scripts\", \"utils.r\"), envir = utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e3b4a406-ffaf-4d20-84f4-893e26dd7217",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[*] Processing: Sig_DEG_High infection_vs_No infection_LFC0.58\n",
      "\n",
      "Warning message in clusterProfiler::bitr(symbols, fromType = \"SYMBOL\", toType = \"ENTREZID\", :\n",
      "\"5.54% of input gene IDs are fail to map...\"\n",
      "Mapped 921 genes to Entrez IDs.\n",
      "\n",
      "[+] Results saved to: ./Pesudobulk/GO_results/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "utils$run_go_enrichment(\"./Pseudobulk/Sig_deg/Sig_DEG_High infection_vs_No infection_LFC0.58.csv\",plot_title = \"High-infection vs No-infection\", out_dir_csv = \"./Pseudobulk/GO_results/\",\n",
    "                              out_dir_fig = \"./Pseudobulk/GO_plots/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7cf7a598-466a-4c4c-902d-ea9e8e04efe3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[*] Processing: Sig_DEG_Low infection_vs_No infection_LFC0.58\n",
      "\n",
      "Warning message in clusterProfiler::bitr(symbols, fromType = \"SYMBOL\", toType = \"ENTREZID\", :\n",
      "\"4.97% of input gene IDs are fail to map...\"\n",
      "Mapped 727 genes to Entrez IDs.\n",
      "\n",
      "[+] Results saved to: ./Pesudobulk/GO_results/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "utils$run_go_enrichment(\"./Pseudobulk/Sig_deg/Sig_DEG_Low infection_vs_No infection_LFC0.58.csv\",plot_title = \"Low-infection vs No-infection\", out_dir_csv = \"./Pseudobulk/GO_results/\",\n",
    "                              out_dir_fig = \"./Pseudobulk/GO_plots/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b719e7e4-f5b3-4424-97a0-b525e6949f37",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[*] Processing: Sig_DEG_High infection_vs_Low infection_LFC0.58\n",
      "\n",
      "Warning message in clusterProfiler::bitr(symbols, fromType = \"SYMBOL\", toType = \"ENTREZID\", :\n",
      "\"3.88% of input gene IDs are fail to map...\"\n",
      "Mapped 421 genes to Entrez IDs.\n",
      "\n",
      "[+] Results saved to: ./Pesudobulk/GO_results/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "utils$run_go_enrichment(\"./Pseudobulk/Sig_deg/Sig_DEG_High infection_vs_Low infection_LFC0.58.csv\",plot_title = \"High-infection vs Low-infection\", out_dir_csv = \"./Pseudobulk/GO_results/\",\n",
    "                              out_dir_fig = \"./Pseudobulk/GO_plots/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0d58bf18-78eb-4306-9945-15634b388bde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.2.3 (2023-03-15 ucrt)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 26200)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=Korean_Korea.utf8  LC_CTYPE=Korean_Korea.utf8   \n",
       "[3] LC_MONETARY=Korean_Korea.utf8 LC_NUMERIC=C                 \n",
       "[5] LC_TIME=Korean_Korea.utf8    \n",
       "\n",
       "attached base packages:\n",
       "[1] tools     stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] here_1.0.2            ggplot2_3.5.1         enrichplot_1.18.4    \n",
       " [4] clusterProfiler_4.6.2 org.Hs.eg.db_3.16.0   AnnotationDbi_1.60.2 \n",
       " [7] IRanges_2.32.0        S4Vectors_0.36.2      Biobase_2.58.0       \n",
       "[10] BiocGenerics_0.44.0  \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] fgsea_1.24.0           colorspace_2.1-0       ggtree_3.6.2          \n",
       "  [4] gson_0.1.0             rprojroot_2.1.1        IRdisplay_1.1         \n",
       "  [7] qvalue_2.30.0          XVector_0.38.0         base64enc_0.1-3       \n",
       " [10] fs_1.6.3               aplot_0.2.9            farver_2.1.1          \n",
       " [13] graphlayouts_1.1.1     ggrepel_0.9.5          bit64_4.0.5           \n",
       " [16] fansi_1.0.6            scatterpie_0.2.6       codetools_0.2-20      \n",
       " [19] splines_4.2.3          cachem_1.0.8           GOSemSim_2.24.0       \n",
       " [22] polyclip_1.10-6        IRkernel_1.3.2         jsonlite_1.8.8        \n",
       " [25] GO.db_3.16.0           png_0.1-8              ggforce_0.4.2         \n",
       " [28] compiler_4.2.3         httr_1.4.7             Matrix_1.5-3          \n",
       " [31] fastmap_1.1.1          lazyeval_0.2.2         cli_3.6.2             \n",
       " [34] tweenr_2.0.3           htmltools_0.5.8.1      igraph_2.0.3          \n",
       " [37] gtable_0.3.6           glue_1.7.0             GenomeInfoDbData_1.2.9\n",
       " [40] reshape2_1.4.4         dplyr_1.1.4            rappdirs_0.3.3        \n",
       " [43] fastmatch_1.1-4        Rcpp_1.0.12            vctrs_0.6.5           \n",
       " [46] Biostrings_2.66.0      ape_5.8                nlme_3.1-162          \n",
       " [49] ggraph_2.2.1           stringr_1.5.2          lifecycle_1.0.4       \n",
       " [52] DOSE_3.24.2            zlibbioc_1.44.0        MASS_7.3-58.2         \n",
       " [55] scales_1.4.0           tidygraph_1.3.1        ragg_1.3.0            \n",
       " [58] parallel_4.2.3         RColorBrewer_1.1-3     memoise_2.0.1         \n",
       " [61] gridExtra_2.3          downloader_0.4.1       ggfun_0.2.0           \n",
       " [64] HDO.db_0.99.1          yulab.utils_0.2.1      stringi_1.8.3         \n",
       " [67] RSQLite_2.3.6          tidytree_0.4.6         BiocParallel_1.32.6   \n",
       " [70] repr_1.1.7             GenomeInfoDb_1.34.9    rlang_1.1.3           \n",
       " [73] pkgconfig_2.0.3        systemfonts_1.0.6      bitops_1.0-7          \n",
       " [76] ggiraph_0.8.9          evaluate_1.0.5         lattice_0.20-45       \n",
       " [79] purrr_1.0.2            labeling_0.4.3         treeio_1.22.0         \n",
       " [82] patchwork_1.2.0        htmlwidgets_1.6.4      cowplot_1.1.3         \n",
       " [85] shadowtext_0.1.6       bit_4.0.5              tidyselect_1.2.1      \n",
       " [88] plyr_1.8.9             magrittr_2.0.3         R6_2.6.1              \n",
       " [91] generics_0.1.4         pbdZMQ_0.3-11          DBI_1.2.3             \n",
       " [94] pillar_1.9.0           withr_3.0.2            KEGGREST_1.38.0       \n",
       " [97] RCurl_1.98-1.14        tibble_3.2.1           crayon_1.5.3          \n",
       "[100] uuid_1.2-0             utf8_1.2.4             viridis_0.6.5         \n",
       "[103] grid_4.2.3             data.table_1.15.4      blob_1.2.4            \n",
       "[106] S7_0.1.1               digest_0.6.35          tidyr_1.3.1           \n",
       "[109] gridGraphics_0.5-1     textshaping_0.3.7      viridisLite_0.4.2     \n",
       "[112] ggplotify_0.1.3       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
