{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76f08eeb-8dc8-4a7a-85a7-4da74c65525b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current work directory: C:\\Users\\yyyyx\\Work\\OC43_machine_learning\\Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "if os.getcwd().endswith(\"notebooks\"):\n",
    "    os.chdir(\"..\")\n",
    "\n",
    "sys.path.append(os.path.abspath(\"./\"))\n",
    "\n",
    "print(f\"Current work directory: {os.getcwd()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6f1ff80d-0707-46a8-9e86-c6ed26a304d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.sparse as sp\n",
    "import gseapy as gp\n",
    "import scripts.utils as ut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8a10edd0-5362-40a8-a055-de252dc201f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('./Data/h5ad/Visualize.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7dc4d3ce-19a8-4513-8806-6f00a5eb3209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12980 × 1614\n",
       "    obs: 'condition', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_oc43', 'pct_counts_oc43', 'viral_fraction', 'viral_per_10k', 'y_log1p_per10k', 'n_genes', 'leiden', 'infection_group'\n",
       "    var: 'gene_ids', 'feature_types', 'mt', 'oc43', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells'\n",
       "    uns: 'leiden', 'log1p', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bec1e15f-0184-4bcb-b4d2-f1a61c270c2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating pseudobulk for groups in: infection_group\n",
      "Completed: 9 samples, 1614 genes.\n"
     ]
    }
   ],
   "source": [
    "pb_counts, pb_meta = ut.generate_pseudoreplicate_bulk(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "56b70fdc-61a9-4777-ad36-2dcfe74aa4f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>sum</th>\n",
       "      <th>median</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>infection_group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>High infection</th>\n",
       "      <td>3</td>\n",
       "      <td>3156</td>\n",
       "      <td>1052.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Low infection</th>\n",
       "      <td>3</td>\n",
       "      <td>4102</td>\n",
       "      <td>1367.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>No infection</th>\n",
       "      <td>3</td>\n",
       "      <td>5722</td>\n",
       "      <td>1907.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 count   sum  median\n",
       "infection_group                     \n",
       "High infection       3  3156  1052.0\n",
       "Low infection        3  4102  1367.0\n",
       "No infection         3  5722  1907.0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pb_meta.groupby('infection_group')['n_cells'].agg(['count','sum','median'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cb9f5143-8713-40b0-8e49-7dafcec7cfb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pb_counts.to_csv('./CSV/pb_counts.csv')\n",
    "pb_meta.to_csv('./CSV/pb_meta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "916c46fc-78a7-4614-9867-01383853a8cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "anndata             0.11.3\n",
       "gseapy              1.1.10\n",
       "matplotlib          3.10.0\n",
       "numpy               2.1.3\n",
       "pandas              3.0.0\n",
       "scanpy              1.10.4\n",
       "scipy               1.15.1\n",
       "scripts             NA\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                 11.1.0\n",
       "asttokens           NA\n",
       "bottleneck          1.4.2\n",
       "brotlicffi          1.0.9.2\n",
       "certifi             2025.10.05\n",
       "charset_normalizer  3.3.2\n",
       "colorama            0.4.6\n",
       "comm                0.2.1\n",
       "cycler              0.12.1\n",
       "cython_runtime      NA\n",
       "dateutil            2.9.0.post0\n",
       "debugpy             1.8.11\n",
       "decorator           5.1.1\n",
       "executing           0.8.3\n",
       "h5py                3.12.1\n",
       "idna                3.7\n",
       "igraph              0.11.5\n",
       "ipykernel           6.29.5\n",
       "jedi                0.19.2\n",
       "joblib              1.4.2\n",
       "kiwisolver          1.4.7\n",
       "legacy_api_wrap     NA\n",
       "leidenalg           0.10.2\n",
       "llvmlite            0.44.0\n",
       "mpl_toolkits        NA\n",
       "natsort             8.4.0\n",
       "nt                  NA\n",
       "numba               0.61.0\n",
       "numexpr             2.11.0\n",
       "packaging           24.2\n",
       "parso               0.8.4\n",
       "platformdirs        3.10.0\n",
       "prompt_toolkit      3.0.43\n",
       "psutil              5.9.0\n",
       "pure_eval           0.2.2\n",
       "pydev_ipython       NA\n",
       "pydevconsole        NA\n",
       "pydevd              3.2.3\n",
       "pydevd_file_utils   NA\n",
       "pydevd_plugins      NA\n",
       "pydevd_tracing      NA\n",
       "pygments            2.15.1\n",
       "pyparsing           3.2.1\n",
       "pythoncom           NA\n",
       "pytz                2024.1\n",
       "pywin32_system32    NA\n",
       "pywintypes          NA\n",
       "requests            2.32.5\n",
       "six                 1.17.0\n",
       "sklearn             1.6.1\n",
       "socks               1.7.1\n",
       "stack_data          0.2.0\n",
       "texttable           1.7.0\n",
       "threadpoolctl       3.5.0\n",
       "tornado             6.4.2\n",
       "traitlets           5.14.3\n",
       "typing_extensions   NA\n",
       "urllib3             2.5.0\n",
       "wcwidth             0.2.5\n",
       "win32api            NA\n",
       "win32com            NA\n",
       "zmq                 26.2.0\n",
       "zoneinfo            NA\n",
       "zope                NA\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.30.0\n",
       "jupyter_client      8.6.3\n",
       "jupyter_core        5.7.2\n",
       "-----\n",
       "Python 3.11.11 | packaged by Anaconda, Inc. | (main, Dec 11 2024, 16:34:19) [MSC v.1929 64 bit (AMD64)]\n",
       "Windows-10-10.0.26200-SP0\n",
       "-----\n",
       "Session information updated at 2026-02-13 15:44\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "\n",
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
