{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "763f02aa-0344-47f8-8f3e-4e67ac425667",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# =========================================\n",
    "# DESeq2 for pseudo-replicate pseudo-bulk\n",
    "# =========================================\n",
    "\n",
    "suppressPackageStartupMessages({\n",
    "library(DESeq2)\n",
    "library(tidyverse)\n",
    "library(apeglm)\n",
    "library(here)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "acde6ad1-0ffe-46df-8ed9-dfc21dd0dbf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "034a0d6a-8df0-428a-b6ee-34dbd1852330",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning'"
      ],
      "text/latex": [
       "'C:/Users/yyyyx/Work/OC43\\_machine\\_learning/Regression\\_Modeling\\_of\\_OC43\\_Coronavirus\\_Viral\\_Burden\\_from\\_Host\\_Single-cell\\_Transcriptomes\\_Using\\_Explainable\\_Machine\\_Learning'"
      ],
      "text/markdown": [
       "'C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning'"
      ],
      "text/plain": [
       "[1] \"C:/Users/yyyyx/Work/OC43_machine_learning/Regression_Modeling_of_OC43_Coronavirus_Viral_Burden_from_Host_Single-cell_Transcriptomes_Using_Explainable_Machine_Learning\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5441cbb1-5b28-4395-ad66-a49199e18490",
   "metadata": {},
   "outputs": [],
   "source": [
    "utils <- new.env()\n",
    "\n",
    "sys.source(here::here(\"scripts\", \"utils.r\"), envir = utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e525ac9e-3ffb-44b4-9fbc-a741ff99fba6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mNew names:\n",
      "\u001b[36m•\u001b[39m `` -> `...1`\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m1614\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (1): ...1\n",
      "\u001b[32mdbl\u001b[39m (9): No infection_rep1, No infection_rep2, No infection_rep3, Low infect...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m9\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m5\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (2): pseudo_sample, infection_group\n",
      "\u001b[32mdbl\u001b[39m (3): replicate, n_cells, umi_sum\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "counts <- read_csv('./CSV/pb_counts.csv') %>%  as.data.frame()\n",
    "coldata <- read_csv(\"./CSV/pb_meta.csv\") %>%  as.data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "76dcb17f-ab64-4ed4-b006-fe6e212f61a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting rownames\n",
    "rownames(counts) <- counts[[1]]; counts[[1]] <- NULL\n",
    "rownames(coldata) <- coldata[[1]]; coldata[[1]] <- NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "944d29cf-188c-47e8-988f-8e36e009bbad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "\"some variables in design formula are characters, converting to factors\"\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeqDataSetFromMatrix(countData = counts,\n",
    "                              colData = coldata,\n",
    "                              design = ~ infection_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0cd805e7-81f6-419a-b784-61a633eb6be9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeq(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4ef161e4-33ae-4cfe-95fa-b4e6c1fec557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: DESeqDataSet \n",
       "dim: 1614 9 \n",
       "metadata(1): version\n",
       "assays(4): counts mu H cooks\n",
       "rownames(1614): SAMD11 HES4 ... MT-ND5 MT-ND6\n",
       "rowData names(26): baseMean baseVar ... deviance maxCooks\n",
       "colnames(9): No infection_rep1 No infection_rep2 ... High\n",
       "  infection_rep2 High infection_rep3\n",
       "colData names(5): infection_group replicate n_cells umi_sum sizeFactor"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8abaf6f5-ded3-45d8-a026-5d7fee94567d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Intercept'</li><li>'infection_group_Low.infection_vs_High.infection'</li><li>'infection_group_No.infection_vs_High.infection'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Intercept'\n",
       "\\item 'infection\\_group\\_Low.infection\\_vs\\_High.infection'\n",
       "\\item 'infection\\_group\\_No.infection\\_vs\\_High.infection'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Intercept'\n",
       "2. 'infection_group_Low.infection_vs_High.infection'\n",
       "3. 'infection_group_No.infection_vs_High.infection'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"Intercept\"                                      \n",
       "[2] \"infection_group_Low.infection_vs_High.infection\"\n",
       "[3] \"infection_group_No.infection_vs_High.infection\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resultsNames(dds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f40f320f-2255-4c4e-bca6-726302a67ee1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing: Low infection versus No infection ...\n",
      "\n",
      "Warning message in write.csv(res_df, file = paste0(save_path, \"DEG_results_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Warning message in write.csv(sig_deg, file = paste0(save_path, \"Sig_DEG_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Done: Low infection | Found 765 DEGs\n",
      "\n",
      "Processing: High infection versus No infection ...\n",
      "\n",
      "Warning message in write.csv(res_df, file = paste0(save_path, \"DEG_results_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Warning message in write.csv(sig_deg, file = paste0(save_path, \"Sig_DEG_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Done: High infection | Found 975 DEGs\n",
      "\n",
      "Processing: High infection versus Low infection ...\n",
      "\n",
      "Warning message in write.csv(res_df, file = paste0(save_path, \"DEG_results_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Warning message in write.csv(sig_deg, file = paste0(save_path, \"Sig_DEG_\", coefficient, :\n",
      "\"attempt to set 'col.names' ignored\"\n",
      "Done: High infection | Found 438 DEGs\n",
      "\n"
     ]
    }
   ],
   "source": [
    "utils$analyze_save(dds, \"infection_group\", \"Low infection\", \"No infection\", './Pseudobulk/Sig_Deg/', lfc_cutoff = 0.58)\n",
    "utils$analyze_save(dds, \"infection_group\", \"High infection\", \"No infection\", './Pseudobulk/Sig_Deg/', lfc_cutoff = 0.58)\n",
    "utils$analyze_save(dds, \"infection_group\", \"High infection\", \"Low infection\", './Pseudobulk/Sig_Deg/', lfc_cutoff = 0.58)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1f87d1d2-4c1a-4328-802d-b65fbc01651d",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "  library(matrixStats) \n",
    "})\n",
    "\n",
    "# ===== PCA (DESeq2 vst based) =====\n",
    "vsd <- vst(dds, blind = TRUE)\n",
    "mat <- assay(vsd)\n",
    "\n",
    "pca <- prcomp(t(mat), center = TRUE, scale. = FALSE)\n",
    "\n",
    "pca_df <- as.data.frame(pca$x)\n",
    "pca_df$group  <- colData(dds)$condition\n",
    "pca_df$sample <- colnames(mat)\n",
    "\n",
    "var_ratio <- (pca$sdev ^ 2) / sum(pca$sdev ^ 2)\n",
    "pc1_var   <- round(var_ratio[1] * 100, 1)\n",
    "pc2_var   <- round(var_ratio[2] * 100, 1)\n",
    "\n",
    "vr <- data.frame(PC = paste0(\"PC\", seq_along(var_ratio)),\n",
    "                 ratio = var_ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9c9c0f18-81d6-4995-b3ff-1ea903ed765b",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(pca_df, './Pseudobulk/pca.csv', row.names = FALSE)\n",
    "write.csv(vr, \"./Pseudobulk/variance.csv\", row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ab3f606f-0242-4c45-a92b-8a87f301a322",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.2.3 (2023-03-15 ucrt)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 26200)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=Korean_Korea.utf8  LC_CTYPE=Korean_Korea.utf8   \n",
       "[3] LC_MONETARY=Korean_Korea.utf8 LC_NUMERIC=C                 \n",
       "[5] LC_TIME=Korean_Korea.utf8    \n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] here_1.0.2                  apeglm_1.20.0              \n",
       " [3] lubridate_1.9.3             forcats_1.0.1              \n",
       " [5] stringr_1.5.2               dplyr_1.1.4                \n",
       " [7] purrr_1.0.2                 readr_2.1.5                \n",
       " [9] tidyr_1.3.1                 tibble_3.2.1               \n",
       "[11] ggplot2_3.5.1               tidyverse_2.0.0            \n",
       "[13] DESeq2_1.38.3               SummarizedExperiment_1.28.0\n",
       "[15] Biobase_2.58.0              MatrixGenerics_1.10.0      \n",
       "[17] matrixStats_1.3.0           GenomicRanges_1.50.2       \n",
       "[19] GenomeInfoDb_1.34.9         IRanges_2.32.0             \n",
       "[21] S4Vectors_0.36.2            BiocGenerics_0.44.0        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] bitops_1.0-7           bit64_4.0.5            RColorBrewer_1.1-3    \n",
       " [4] httr_1.4.7             rprojroot_2.1.1        repr_1.1.7            \n",
       " [7] numDeriv_2016.8-1.1    tools_4.2.3            utf8_1.2.4            \n",
       "[10] R6_2.6.1               DBI_1.2.3              colorspace_2.1-0      \n",
       "[13] withr_3.0.2            tidyselect_1.2.1       bit_4.0.5             \n",
       "[16] compiler_4.2.3         textshaping_0.3.7      cli_3.6.2             \n",
       "[19] DelayedArray_0.23.2    scales_1.4.0           mvtnorm_1.2-4         \n",
       "[22] pbdZMQ_0.3-11          systemfonts_1.0.6      digest_0.6.35         \n",
       "[25] XVector_0.38.0         base64enc_0.1-3        pkgconfig_2.0.3       \n",
       "[28] htmltools_0.5.8.1      fastmap_1.1.1          bbmle_1.0.25.1        \n",
       "[31] rlang_1.1.3            RSQLite_2.3.6          farver_2.1.1          \n",
       "[34] generics_0.1.4         jsonlite_1.8.8         vroom_1.6.5           \n",
       "[37] BiocParallel_1.32.6    RCurl_1.98-1.14        magrittr_2.0.3        \n",
       "[40] GenomeInfoDbData_1.2.9 Matrix_1.5-3           Rcpp_1.0.12           \n",
       "[43] IRkernel_1.3.2         fansi_1.0.6            lifecycle_1.0.4       \n",
       "[46] stringi_1.8.3          MASS_7.3-58.2          zlibbioc_1.44.0       \n",
       "[49] plyr_1.8.9             grid_4.2.3             blob_1.2.4            \n",
       "[52] parallel_4.2.3         bdsmatrix_1.3-7        crayon_1.5.3          \n",
       "[55] lattice_0.20-45        IRdisplay_1.1          Biostrings_2.66.0     \n",
       "[58] annotate_1.76.0        hms_1.1.3              KEGGREST_1.38.0       \n",
       "[61] locfit_1.5-9.9         pillar_1.9.0           uuid_1.2-0            \n",
       "[64] geneplotter_1.76.0     codetools_0.2-20       XML_3.99-0.16.1       \n",
       "[67] glue_1.7.0             evaluate_1.0.5         png_0.1-8             \n",
       "[70] vctrs_0.6.5            tzdb_0.4.0             gtable_0.3.6          \n",
       "[73] cachem_1.0.8           emdbook_1.3.14         xtable_1.8-4          \n",
       "[76] coda_0.19-4.1          ragg_1.3.0             AnnotationDbi_1.60.2  \n",
       "[79] memoise_2.0.1          timechange_0.3.0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
